mlir/tools/mlir-rocm-runner/rocm-runtime-wrappers.cpp

```diff
@@ -353,122 +354,81 @@ extern "C" void mgpuMemCopy4DBF16(unsigned short *sourceAllocated, unsigned shor
+  auto filterStrides = llvm::ArrayRef<int64_t>(filter->strides, rank1);
+
+  auto *input = static_cast<StridedMemRefType<float, 4> *>(i_ptr);
+  auto *inputAllocated = input->data + input->offset;
+  auto inputSizes = llvm::ArrayRef<int64_t>(input->sizes, rank2);
+  auto inputStrides = llvm::ArrayRef<int64_t>(input->strides, rank2);
+
+  auto *output = static_cast<StridedMemRefType<float, 4> *>(o_ptr);
+  auto *outputAllocated = output->data + output->offset;
+  auto outputSizes = llvm::ArrayRef<int64_t>(output->sizes, rank3);
+  auto outputStrides = llvm::ArrayRef<int64_t>(output->strides, rank3);
+
+  auto *layout1 = static_cast<StridedMemRefType<char, 1> *>(f_layout);
+  auto *filterLayout = layout1->data + layout1->offset;
+
+  auto *layout2 = static_cast<StridedMemRefType<char, 1> *>(i_layout);
+  auto *inputLayout = layout2->data + layout2->offset;
+
+  auto *layout3 = static_cast<StridedMemRefType<char, 1> *>(o_layout);
+  auto *outputLayout = layout3->data + layout3->offset;
 
   // Extract proper tensor sizes and strides based on layouts
+  std::unordered_map<char, std::pair<int, int>> filterSizeStride;
+  std::unordered_map<char, std::pair<int, int>> inputSizeStride;
+  std::unordered_map<char, std::pair<int, int>> outputSizeStride;
   for (size_t i = 0; i < 4; i++) {
-    if (filterLayout[i] == 'k') {
-      filterOutChanStride = filterStrides[i];
-    } else if (filterLayout[i] == 'c') {
-      filterInChanStride = filterStrides[i];
-    } else if (filterLayout[i] == 'y') {
-      filterHeightSize = filterSizes[i];
-      filterHeightStride = filterStrides[i];
-    } else if (filterLayout[i] == 'x') {
-      filterWidthSize = filterSizes[i];
-      filterWidthStride = filterStrides[i];
-    }
-
-    if (inputLayout[i] == 'n') {
-      inBatchStride = inputStrides[i];
-    } else if (inputLayout[i] == 'c') {
-      inChannelSize = inputSizes[i];
-      inChannelStride = inputStrides[i];
-    } else if (inputLayout[i] == 'h') {
-      inHeightSize = inputSizes[i];
-      inHeightStride = inputStrides[i];
-    } else if (inputLayout[i] == 'w') {
-      inWidthSize = inputSizes[i];
-      inWidthStride = inputStrides[i];
-    }
-
-    if (outputLayout[i] == 'n') {
-      batchSize = outputSizes[i];
-      outBatchStride = outputStrides[i];
-    } else if (outputLayout[i] == 'k') {
-      outChannelSize = outputSizes[i];
-      outChannelStride = outputStrides[i];
-    } else if (outputLayout[i] == 'h') {
-      outHeightSize = outputSizes[i];
-      outHeightStride = outputStrides[i];
-    } else if (outputLayout[i] == 'w') {
-      outWidthSize = outputSizes[i];
-      outWidthStride = outputStrides[i];
-    }
+    filterSizeStride[filterLayout[i]] =
+        std::make_pair(filterSizes[i], filterStrides[i]);
+    inputSizeStride[inputLayout[i]] =
+        std::make_pair(inputSizes[i], inputStrides[i]);
+    outputSizeStride[outputLayout[i]] =
+        std::make_pair(outputSizes[i], outputStrides[i]);
```

Comment:
Looking much cleaner now, thanks!

Though I believe the `StridedMemRefType` use `int64_t` whereas you use `int` which silently downsized the type. This can create a problem when you have larger tensors who's stride exceed integer. I'd advise to change back to `int64_t` for both size and strides.